src/core/providers/ogr/qgsogrprovider.cpp

```diff
@@ -1136,55 +1194,95 @@ QgsRectangle QgsOgrProvider::extent() const
       mOgrOrigLayer->ExecuteSQLNoReturn( sql );
     }
 
-    mExtent->MinX = std::numeric_limits<double>::max();
-    mExtent->MinY = std::numeric_limits<double>::max();
-    mExtent->MaxX = -std::numeric_limits<double>::max();
-    mExtent->MaxY = -std::numeric_limits<double>::max();
-
-    // TODO: This can be expensive, do we really need it!
-    if ( mOgrLayer == mOgrOrigLayer.get() && mSubsetString.isEmpty() )
+    if ( mOgrLayer == mOgrOrigLayer.get() )
     {
-      if ( ( mGDALDriverName == QLatin1String( "OAPIF" ) || mGDALDriverName == QLatin1String( "WFS3" ) ) &&
-           !mOgrLayer->TestCapability( OLCFastGetExtent ) )
+      if ( !mOgrLayer->TestCapability( OLCFastGetExtent ) )
       {
-        // When the extent is not in the metadata, retrieving it would be
-        // super slow
-        mExtent->MinX = -180;
-        mExtent->MinY = -90;
-        mExtent->MaxX = 180;
-        mExtent->MaxY = 90;
+        QgsDebugMsgLevel( QStringLiteral( "WITHOUT OLCFastGetExtent" ), 3 );
+        if ( mGDALDriverName == QLatin1String( "OAPIF" ) || mGDALDriverName == QLatin1String( "WFS3" ) )
+        {
+          QgsDebugMsgLevel( QStringLiteral( "WITHOUT OLCFastGetExtent AND is remote" ), 3 );
+          // When the extent is not in the metadata, retrieving it would be
+          // super slow for remote data sources
+          mExtent->MinX = -180;
+          mExtent->MinY = -90;
+          mExtent->MinZ = std::numeric_limits<double>::quiet_NaN();
+          mExtent->MaxX = 180;
+          mExtent->MaxY = 90;
+          mExtent->MaxZ = std::numeric_limits<double>::quiet_NaN();
+          hasBeenComputed = true;
+        }
       }
       else
       {
-        mOgrLayer->GetExtent( mExtent.get(), true );
+        if ( is3D && mSubsetString.isEmpty() ) // TODO: find a way to use mSubsetString in the request
+        {
+          QgsDebugMsgLevel( QStringLiteral( "WITH OLCFastGetExtent" ), 3 );
+          QRecursiveMutex *mutex = nullptr;
+          OGRLayerH layer = mOgrLayer->getHandleAndMutex( mutex );
+          const char *geomCol = OGR_L_GetGeometryColumn( layer );
+          if ( geomCol != NULL && strlen( geomCol ) > 0 )
+          {
+            QgsDebugMsgLevel( QStringLiteral( "WITH OLCFastGetExtent AND WITH geomCol: %1" ).arg( geomCol ), 3 );
+
+            mOgrLayer->GetExtent( mExtent.get(), true );
+            QByteArray geomColQuoted = quotedIdentifier( geomCol );
+            QByteArray sql = "SELECT MIN(ST_MinZ("
+                             + geomColQuoted
+                             + ")), MAX(ST_MaxZ("
+                             + geomColQuoted
+                             + ")) FROM "
+                             + quotedIdentifier( mOgrLayer->name() );
```

Comment:
this will only work with PostGIS, GPKG and Spatialite datasets. You probably want to run the ExecuteSQL() between CPLPushErrorHandler(CPLQuietErrorHandler) / CPLPopErrorHandler() to avoid error messages on datasets that don't support that

Reply:
> The check `if ( l ) {...}` will be enough?

yes